_posts/2023-05-10-group-equivariant-cnn-3/group-equivariant-cnn-3.Rmd

```diff
@@ -0,0 +1,492 @@
+
+First, to map my understanding of the mathematical concepts to the terminology used in the library, and from there, to correct use and application. Expressed schematically: We have a concept A, which figures (among other concepts) in technical term (or object class) B. What does my understanding of A tell me about how object class B is to be used correctly? More importantly: How do I use it to best attain my goal C? -- This first difficulty I'll address in a very pragmatic way. I'll neither dwell on mathematical details, nor try to establish the links between A, B, and C in detail. Instead, I'll present the characters[^2] in this story by asking what they're good for.
+
+[^2]: If you have some background on representations: No, not *those* characters ...
+
+Second -- and this will be of relevance to just a subset of readers -- the topic of group equivariance, particularly as applied to image processing, is one where visualizations can be of tremendous help. The quaternity[^3] of conceptual explanation, math, code, and visualization can, together, produce an understanding of emergent-seeming quality... if, and only if, all of these explanation modes "work" for you. (Or if, in an area, a mode that does not wouldn't contribute that much anyway.) Here, it so happens that from what I saw, several papers have excellent visualizations[^4], and the same holds for some lecture slides and accompanying notebooks[^5]. But for those among us with limited spatial-imagination capabilities -- e.g., people with [Aphantasia](https://en.wikipedia.org/wiki/Aphantasia) -- these illustrations, intended to help, can be very hard to make sense of themselves. If you're not one of these, I totally recommend checking out the resources linked in the above footnotes. This text, though, will try to make the best possible use of verbal explanation to introduce the concepts involved, the library, and how to use it.
+
+[^3]: Yes, that word [exists](https://www.merriam-webster.com/dictionary/quaternity), although I must admit I didn't know before typing it into a search engine.
+
+[^4]: One paper particularly stood out to me: [@abs-2106-06020].
+
+[^5]: Directly pertinent to today's topic, thinking of the materials produced for University of Amsterdam's [course on group-equivariant deep learning](https://uvagedl.github.io/).
+
+That said, let's start with the software.
+
+## Using *escnn*
+
+`Escnn` depends on PyTorch. Yes, PyTorch, not `torch`; unfortunately, the library hasn't been ported to R yet. (If, after reading this post, you feel that maybe *you* would be interested in porting it -- I can definitely say I think that's a great idea!) For now, thus, we'll employ [`reticulate`](https://rstudio.github.io/reticulate/)[^6] to access the Python objects directly.
```

Comment:
> (If, after reading this post, you feel that maybe *you* would be interested in porting it -- I can definitely say I think that's a great idea!)

This could maybe be a footnote rather than a parenthetical, since you're starting to really jump into the explanation now.

Reply:
Oh, true, will do! Too much of a "sales pitch" (or the opposite, sort of)